### ISSUE TITLE
Failing to deploy services due to "helm dependency build" failure

### ISSUE THREAD

**Issue by zonnie** [argoproj/argo-cd#5107]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Some `Application`s based on `helm` fail to deploy due to somekind of internal filesystem issue.

For example, one of the apps that are in `Unknown` states
```
Name:               redis-rule-hit-count-buffer-test-2
Project:            default
Server:             https://kubernetes.default.svc
Namespace:          test-2
URL:                https://35.232.222.64/applications/redis-rule-hit-count-buffer-test-2
Repo:               git@github.com:gc-org/gc-saas-prod.git
Target:             HEAD
Path:               prod/common/infra/redis
SyncWindow:         Sync Allowed
Sync Policy:        Automated (Prune)
Sync Status:        Unknown
Health Status:      Healthy

CONDITION        MESSAGE                                                                                                                                                                                                                                                  LAST TRANSITION
ComparisonError  rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists  2020-12-22 10:14:56 +0200 IST
```
This doesn't eventually resolve itself, it's stays this way...

**To Reproduce**

I'm not sure how to reproduce, this happens from time to time and causes complete deadlock

**Expected behavior**

The `Application` should be deployed successfully

**Screenshots**

![Screen Shot 2020-12-22 at 10 37 37](https://user-images.githubusercontent.com/7836993/102867641-baa6f900-4441-11eb-9fe8-5a45edfd7c29.png)

**Version**

```shell
argocd: v1.7.10+bcb05b0.dirty
  BuildDate: 2020-11-21T00:34:29Z
  GitCommit: bcb05b0c2e0f8006aa2d2abaf780e73c9e73c945
  GitTreeState: dirty
  GoVersion: go1.15.5
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.8.1+c2547dc
  BuildDate: 2020-12-10T02:59:21Z
  GitCommit: c2547dca95437fdbb4d1e984b0592e6b9110d37f
  GitTreeState: clean
  GoVersion: go1.14.12
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v3.8.1 2020-07-16T00:58:46Z
  Helm Version: v3.4.1+gc4e7485
  Kubectl Version: v1.17.8
```

**Logs**

```
time="2020-12-22T08:14:58Z" level=info msg="Normalized app spec: {\"status\":{\"conditions\":[{\"lastTransitionTime\":\"2020-12-22T08:14:56Z\",\"message\":\"rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists\",\"type\":\"ComparisonError\"}]}}" application=redis-rule-hit-count-buffer-test-2
```

**Comment by lcostea** [comment]
Might be because of this: https://helm.sh/blog/new-location-stable-incubator-charts/ - the address for stable repo has changed.
@alexmt maybe it would be good to have a 1.7.12 with helm 2.17 included, that has the new address of stable updated in the binary, otherwise upgrade to 1.8 is mandatory

**Comment by JasP19** [comment]
I believe another fix for this may be to update your stable repo in the argocd-cm ConfigMap as follows:
```yaml
data:
  helm.repositories: |
    - url: https://charts.helm.sh/stable
      name: stable
```

After this, refresh your Argo applications.

**Comment by lcostea** [comment]
@zonnie sorry, now I saw you already are on 1.8.1 server, so it is not that

**Comment by lcostea** [comment]
@JasP19 does it work for helm 2? It has the stable repo embedded the article above says you need to remove stable first

**Comment by JasP19** [comment]
@lcostea it worked for us with Helm 2 and ArgoCD v1.6.1

**Comment by jessesuen** [comment]
Is there an example git location which can reproduce this easily? It's hard to fix this without being able to reproduce it.

**Comment by zonnie** [comment]
I'm not sure when this reproduces.
If you can explain what causes the symptom I might be able to try and reproduce.
The error is completely internal and means nothing to me - maybe more details can help me find a way.

**Comment by JasP19** [comment]
> 
> 
> I believe another fix for this may be to update your stable repo in the argocd-cm ConfigMap as follows:
> 
> ```yaml
> data:
>   helm.repositories: |
>     - url: https://charts.helm.sh/stable
>       name: stable
> ```
> 
> After this, refresh your Argo applications.

@zonnie did you get a chance to test the solution I proposed above?

**Comment by zonnie** [comment]
No, I didn't - can u please explain briefly how this is related?

I'll check this solution anyway and report back

**Comment by JasP19** [comment]
> 
> 
> Might be because of this: https://helm.sh/blog/new-location-stable-incubator-charts/ - the address for stable repo has changed.
> @alexmt maybe it would be good to have a 1.7.12 with helm 2.17 included, that has the new address of stable updated in the binary, otherwise upgrade to 1.8 is mandatory

@zonnie this is related to the fact that Helm recently changed the location of the stable repo as mentioned by @lcostea, above. I believe the Helm 2 binary is now failing to resolve/access the old stable repo. By updating the ConfigMap, you override the stable repo to point at the new URL with the new location. We were having the same issue in our cluster. We made this change and noted that it solved the issue when we refreshed our apps.

**Comment by zonnie** [comment]
> > 
> 
> > 
> 
> > Might be because of this: https://helm.sh/blog/new-location-stable-incubator-charts/ - the address for stable repo has changed.
> 
> > @alexmt maybe it would be good to have a 1.7.12 with helm 2.17 included, that has the new address of stable updated in the binary, otherwise upgrade to 1.8 is mandatory
> 
> 
> 
> @zonnie this is related to the fact that Helm recently changed the location of the stable repo as mentioned by @lcostea, above. I believe the Helm 2 binary is now failing to resolve/access the old stable repo. By updating the ConfigMap, you override the stable repo to point at the new URL with the new location. We were having the same issue in our cluster. We made this change and noted that it solved the issue when we refreshed our apps.

Thanks @JasP19 I will give it a shot as I can

**Comment by gzur** [comment]
Is 1.7.11 also affected by this?

**Comment by JasP19** [comment]
@gzur unfortunately I am not certain. If you are using Helm 2 and have suddenly started seeing apps in an `Unkown` state with a `helm dependency build` failure then it may be worth attempting the solution I proposed. It worked for us on ArgoCD v1.6.1 and it's only a minor change.

**Comment by gzur** [comment]
> @gzur unfortunately I am not certain. If you are using Helm 2 and have suddenly started seeing apps in an `Unkown` state with a `helm dependency build` failure then it may be worth attempting the solution I proposed. It worked for us on ArgoCD v1.6.1 and it's only a minor change.

I'm running v1.7.6 as is - I'm going to give updating to 1.7.11 the old college try and report back.

**Comment by gzur** [comment]
> I'm running v1.7.6 as is - I'm going to give updating to 1.7.11 the old college try and report back.

I was able to upgrade to 1.7.11 - but it had no effect.

I ended up applying the following configuration through a custom `values.yaml` file:
```
argocd:
  server:
    config:
      repositories: |
        [...]
        - type: helm
          name: stable
          url: https://charts.helm.sh/stable
        [...]
```
After this, I had to `helm template [...] | kubectl apply [...]` for it to kick in, since ArgoCD is non-functional.

**Comment by gzur** [comment]
> Might be because of this: https://helm.sh/blog/new-location-stable-incubator-charts/ - the address for stable repo has changed.

I believe the error described in this issue is unrelated to the stable repo move. I experienced that first hand, and the error message produced looks like this:
```
helm2 dependency build` failed exit status 1: 
Error: open /tmp/helm769321715/repository/cache/stable-index.yaml: no such file or directory
```
Which I successfully addressed with a configuration change described above: https://github.com/argoproj/argo-cd/issues/5107#issuecomment-751783856

The error described in the issue: 
```
 rename charts tmpcharts: file exists
 ```
 is some helm-related issue (that I actually remember seeing before, but I can't remember what caused it)

**Comment by alex1989hu** [comment]
No matter which version `v1.8.1` or `v1.8.2` both produces the same. Even with option which is provided above / to add `stable` as Helm repository/. Several days passed, sometimes I see 4 `Unknown` state, sometimes ~36 `Unknown` state.

**Comment by ArgTang** [comment]
Same issue here. seems to be a helm issue: https://github.com/helm/helm/issues/5567 with corresponding PR
https://github.com/helm/helm/pull/8846

The issue is intermittent, and will fix itself after a hard reset. How does argocd handle refresh errors? to manually do a hard refresh on our deploys is not a good experience. I think refresh errors could use some improvements regardless of the helm fixes

**Comment by gzur** [comment]
We have been seeing this intermittently, with an App going into `Unknown` state with the aforementioned error message. 

I'm wondering whether this has anything to do with the `application-controller` being too hard on the `reposerver`? 

The `application-controller` is invoking this `GenerateManifest` method over `grpc`, with the `reposerver` acting on that request by interfacing **directly with the file-system** - doing `helm dep update` and such. 

The ArgoCD Application that is causing us the most trouble has **20** Helm charts as as dependencies, and it can take a while for them all to download. Is it possible that the `application-controller` is getting fed up with waiting and retrying, causing the `reposerver` to start running `helm dep update` with another already running? 

This could cause these `./tmpcharts` shenanigans, because apparently, Helm stores the charts in `./tmpcharts` before moving them all over to `./charts` once it has finished fetching all the dependencies.

**Comment by gzur** [comment]
Would setting [`ARGOCD_SYNC_WAVE_DELAY`](https://github.com/argoproj/argo-cd/blob/master/controller/sync.go#L34) to something higher than 2 seconds (the default) affect this at all?

https://github.com/argoproj/argo-cd/blob/master/controller/sync.go#L34

**Comment by gzur** [comment]
> Would setting [`ARGOCD_SYNC_WAVE_DELAY`](https://github.com/argoproj/argo-cd/blob/master/controller/sync.go#L34) to something higher than 2 seconds (the default) affect this at all?

Nope.

**Comment by meshack-mbuvi** [comment]
> > I believe another fix for this may be to update your stable repo in the argocd-cm ConfigMap as follows:
> > ```yaml
> > data:
> >   helm.repositories: |
> >     - url: https://charts.helm.sh/stable
> >       name: stable
> > ```
> > 
> > 
> > After this, refresh your Argo applications.
> 
> @zonnie did you get a chance to test the solution I proposed above?



> > I'm running v1.7.6 as is - I'm going to give updating to 1.7.11 the old college try and report back.
> 
> I was able to upgrade to 1.7.11 - but it had no effect.
> 
> I ended up applying the following configuration through a custom `values.yaml` file:
> 
> ```
> argocd:
>   server:
>     config:
>       repositories: |
>         [...]
>         - type: helm
>           name: stable
>           url: https://charts.helm.sh/stable
>         [...]
> ```
> 
> After this, I had to `helm template [...] | kubectl apply [...]` for it to kick in, since ArgoCD is non-functional.

I am experiencing the same issue on aws eks. I have tried this suggestion but it does not fix the issue.

**Comment by giorgian** [comment]
We too are experiencing this quite often. Our workaround is to just run `redis-cli flushall` in the Redis pod.

**Comment by gzur** [comment]
We finally resolved this by increasing 
* `--repo-server-timeout-seconds` on the application-controller
* `ARGOCD_EXEC_TIMEOUT` on the repo-server

```
argocd:
  controller:
    extraArgs:
      - --repo-server-timeout-seconds
      - "500" 
  repoServer:
    env:
      - name: "ARGOCD_EXEC_TIMEOUT"
        value: "5m"
```

Doesn't make me happy, but until `helm dep update` is thread-safe(https://github.com/helm/helm/pull/8846#issuecomment-768479847), it'll have to do.

**Comment by mylesagray** [comment]
Just popping in to say that I was seeing this, and the `context deadline exceeded` on a Raspberry Pi cluster, Argo instance is here if you're curious as to the config: http://argocd.apps.blah.cloud

What was visible in Grafana was that the repo server pod was consuming all of the (admittedly small) host's CPU, so a combination of the following cleared all of that up for me:

```
controller:
    extraArgs:
    - --repo-server-timeout-seconds
    - "500"
    - --status-processors
    - "10"
    - --operation-processors
    - "5"
repoServer:
  replicas: 3
  env:
    - name: "ARGOCD_EXEC_TIMEOUT"
      value: "5m"
```

The number of replicas on repoServer helped to distribute the load across the K8s cluster nodes, so it wasn't hammering a single node, and the timeout extensions also helped in the case of long running tasks (due to slow CPUs).

**Comment by kshamajain99** [comment]
@zonnie Can you try the solution suggested by @gzur https://github.com/argoproj/argo-cd/issues/5107#issuecomment-776300229 to see if it resolves the issue for you as well?

**Comment by zonnie** [comment]
> @zonnie Can you try the solution suggested by @gzur https://github.com/argoproj/argo-cd/issues/5107#issuecomment-776300229 to see if it resolves the issue for you as well?

I didn't run into this lately so I can't confirm this 😕

**Comment by mattbennett** [comment]
I experienced this under a different situation --

I was changing the target revision of the app in the UI, from a revision where the chart did not exist to one where it did. When the error occurred the UI reverted the change to the target revision

Rather than using the UI, I edited the target revision of the app directly in the manifest. This time the change persisted, and I was able to see a different error returned by `helm template` when ArgoCD tried to render the chart (in my case it was caused by a helm2 parent chart trying to use a helm3 chart as a dependency)

**Comment by ramanNarasimhan77** [comment]
I faced this issue too. In my case, I was trying to deploy a chart from a private helm registry. 

To determine the ideal resource (limits and requests), I had set the minimum in my argocd-ha installation (setup using community maintained helm chart) and planned to keep increasing it gradually until I figure out the right configuration where everything just works.

I observed from the output of kubectl top pods, that RepoServer had almost reached the limits I had set, and hence was taking a lot of time to finish `helm dep update`

After ~90s, helm dependency build timed out in reposerver and  I could see following log

```    
time="2021-02-26T10:54:14Z" level=error msg="`helm dependency build` failed timeout after 1m30s" execID=SVEuN
```

Argo Server log had this error:

```
021/02/26 10:52:00 proto: tag has too few fields: "-"

time="2021-02-26T10:52:00Z" level=info msg="received unary call /application.ApplicationService/Create" grpc.method=Create grpc.request.claims="{\"iat\":1614336702,\"iss\":\"argocd\",\"nbf\":1614336702,\"sub\":\"admin\"}" grpc.request.content="%!v(PANIC=String method: reflect.Value.Interface: cannot return value obtained from unexported field or method)" grpc.service=application.ApplicationService grpc.start_time="2021-02-26T10:52:00Z" span.kind=server system=grpc

time="2021-02-26T10:53:00Z" level=info msg="finished unary call with code InvalidArgument" error="rpc error: code = InvalidArgument desc = application spec is invalid: InvalidSpecError: Unable to generate manifests in <my-app>: rpc error: code = Canceled desc = context canceled" grpc.code=InvalidArgument grpc.method=Create grpc.service=application.ApplicationService grpc.start_time="2021-02-26T10:52:00Z" grpc.time_ms=59999.926 span.kind=server system=grpc
```

I tried increasing the Resources set in Argocd repo server and then the same helm chart got rendered correctly

**Comment by KlavsKlavsen** [comment]
I have the exact same issue - with argocd 1.8.4.
It happened after having tried a few options on metrics-server which crashed all the time..
"rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists"

**Comment by KlavsKlavsen** [comment]
I am pointing to stable charts repo as metrics-server has not been migrated to kubernetes/metrics-server yet

**Comment by KlavsKlavsen** [comment]
I get same error from my velero chart and that is NOT using the deprecated helm stable chart repo.. so it does not seem to be related to the repo source..

**Comment by KlavsKlavsen** [comment]
this is on an aws cluster - and currently no resource limits set. I am trying setting repo-server replica's to 2 instead of 1.

**Comment by KlavsKlavsen** [comment]
It happened again - and now with 2 reposervers.. so it was not enough to scale that up :(
I tried kill'ing both reposerver pods so they got recreated.. but refresh is NOT making it retry.. Error "rename charts tmpcharts: file exists" is not being updated (still X minutes old and getting older). 
After killing 'server' pod also - i could suddenly choose "hard refresh" (is that just hard to hit exactly, or ?) and issue was resolved. (for now).

**Comment by elvis-cai** [comment]
same issue in v2.0.0, after I delete helm cart and add again, got following error msg:
```
rpc error: code = Unknown desc = Manifest generation error (cached): `helm repo add --username ****** --password ****** airwallex-charts https://xxx` failed exit status 1: Error: repository name (xxx-charts) already exists, please specify a different name

**Comment by KlavsKlavsen** [comment]
Just hit this in v2.0.1 also.. when doing recovery test.. so syncing many (10+) applications (all helm applications) reproduces it.

**Comment by mylesagray** [comment]
I too keep hitting this when syncing a large number of applications (29 - but a few of those are app-of-apps) as part of cluster bootstrap.

However, the main problem we are seeing is that once an application enters the Error state with (for example):

```
ComparisonError: rpc error: code = Unknown desc = `helm repo add charts.helm.sh https://charts.helm.sh/stable` failed signal: segmentation fault (core dumped): qemu: uncaught target signal 11 (Segmentation fault) - core dumped
```

Then it will stay with "Current Sync Status" as `Unknown` perpetually, and never retry.

<img width="1230" alt="image" src="https://user-images.githubusercontent.com/609356/128561414-26fe9eff-a8f2-4609-9f78-0026a9a1a559.png">

It would appear there is some digging to be done in the logs to figure out why these apps never retry.

**Comment by KlavsKlavsen** [comment]
Seeing as this issue is "resolved" temporarily - by killing repo-server pods, so they get recreated - its clearly a caching problem.
My guess is that reposerver runs Helm in parallel on SAME pod - which Helm does not actually support - and hence once you hit 2 simultaneous Helm runs on same pod - Helm breaks and never recovers from that (due to caching).
A simple "lock" around Helm - ensuring it is not run simultaenously, should confirm this as the issue (and resolve it :)

**Comment by denstorti** [comment]
It seems there is some mitigation for tmpcharts clean up in `Helm 3.7.0`, https://github.com/helm/helm/pull/9889

Related to this comment.
> https://github.com/argoproj/argo-cd/issues/5107#issuecomment-776300229

It is not as a good solution as having tmp dirs suffixed with a random string, but it will help. 
Probably we should release a version with Helm 3.7.0.

**Comment by zonnie** [comment]
> We finally resolved this by increasing
> 
> * `--repo-server-timeout-seconds` on the application-controller
> * `ARGOCD_EXEC_TIMEOUT` on the repo-server
> 
> ```
> argocd:
>   controller:
>     extraArgs:
>       - --repo-server-timeout-seconds
>       - "500" 
>   repoServer:
>     env:
>       - name: "ARGOCD_EXEC_TIMEOUT"
>         value: "5m"
> ```
> 
> Doesn't make me happy, but until `helm dep update` is thread-safe([helm/helm#8846 (comment)](https://github.com/helm/helm/pull/8846#issuecomment-768479847)), it'll have to do.

@gzur I'm not sure I understand where you pass these values ? can you please elaborate ?

**Comment by gzur** [comment]
I don't have access to these manifests anymore but iirc, I added the command-line argument and environment variable through Helm `values.yaml`:

We were using the ArgoCD chart as a subchart (hence the `argocd:` root key in my example.

The actual configuration points are here:

```
controller:
    extraArgs:
      - --repo-server-timeout-seconds
      - "500" 

```
https://github.com/argoproj/argo-helm/blob/a4b05b43e2f1d7257eb3ece723972dbb340a66ab/charts/argo-cd/values.yaml#L58


```
repoServer:
    env:
      - name: "ARGOCD_EXEC_TIMEOUT"
        value: "5m"

```
https://github.com/argoproj/argo-helm/blob/a4b05b43e2f1d7257eb3ece723972dbb340a66ab/charts/argo-cd/values.yaml#L1227

**Comment by zonnie** [comment]
This is my repo-server
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: argocd
    meta.helm.sh/release-namespace: argocd
  creationTimestamp: "2021-11-09T21:40:57Z"
  generation: 2
  labels:
    app.kubernetes.io/component: repo-server
    app.kubernetes.io/instance: argocd
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: argocd-repo-server
    app.kubernetes.io/part-of: argocd
    app.kubernetes.io/version: v2.1.6
    helm.sh/chart: argo-cd-3.26.5
  managedFields:
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    manager: Go-http-client
    operation: Update
    time: "2021-11-09T21:40:57Z"
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:deployment.kubernetes.io/revision: {}
      f:status:
        f:availableReplicas: {}
        f:conditions:
          .: {}
          k:{"type":"Available"}:
            .: {}
            f:lastTransitionTime: {}
            f:lastUpdateTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
          k:{"type":"Progressing"}:
            .: {}
            f:lastTransitionTime: {}
            f:lastUpdateTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
        f:observedGeneration: {}
        f:readyReplicas: {}
        f:replicas: {}
        f:updatedReplicas: {}
    manager: kube-controller-manager
    operation: Update
    time: "2021-11-09T21:41:44Z"
  name: argocd-repo-server
  namespace: argocd
  resourceVersion: "8271"
  uid: f5d15317-7461-46a1-800a-a6abd0c260b4
spec:
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 5
  selector:
    matchLabels:
      app.kubernetes.io/instance: argocd
      app.kubernetes.io/name: argocd-repo-server
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: repo-server
        app.kubernetes.io/instance: argocd
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: argocd-repo-server
        app.kubernetes.io/part-of: argocd
        app.kubernetes.io/version: v2.1.6
        helm.sh/chart: argo-cd-3.26.5
    spec:
      containers:
      - args:
        - argocd-repo-server
        - --redis
        - argocd-redis-ha-haproxy:6379
        - --logformat
        - text
        - --loglevel
        - info
        env:
        - name: ARGOCD_EXEC_TIMEOUT
          value: 5m
        image: quay.io/argoproj/argocd:v2.1.6
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          tcpSocket:
            port: 8081
          timeoutSeconds: 1
        name: repo-server
        ports:
        - containerPort: 8081
          name: repo-server
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          tcpSocket:
            port: 8081
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /app/config/gpg/source
          name: gpg-keys
        - mountPath: /app/config/gpg/keys
          name: gpg-keyring
        - mountPath: /app/config/ssh
          name: ssh-known-hosts
        - mountPath: /app/config/reposerver/tls
          name: argocd-repo-server-tls
        - mountPath: /tmp
          name: tmp-dir
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: default
      serviceAccountName: default
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          name: argocd-gpg-keys-cm
        name: gpg-keys
      - emptyDir: {}
        name: gpg-keyring
      - configMap:
          defaultMode: 420
          name: argocd-ssh-known-hosts-cm
        name: ssh-known-hosts
      - name: argocd-repo-server-tls
        secret:
          defaultMode: 420
          items:
          - key: tls.crt
            path: tls.crt
          - key: tls.key
            path: tls.key
          - key: ca.crt
            path: ca.crt
          optional: true
          secretName: argocd-repo-server-tls
      - emptyDir: {}
        name: tmp-dir
status:
  availableReplicas: 2
  conditions:
  - lastTransitionTime: "2021-11-09T21:41:44Z"
    lastUpdateTime: "2021-11-09T21:41:44Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2021-11-09T21:40:57Z"
    lastUpdateTime: "2021-11-09T21:41:44Z"
    message: ReplicaSet "argocd-repo-server-6c566b9648" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 2
  readyReplicas: 2
  replicas: 2
  updatedReplicas: 2
```

And my controller 
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  annotations:
    meta.helm.sh/release-name: argocd
    meta.helm.sh/release-namespace: argocd
  creationTimestamp: "2021-11-09T21:40:57Z"
  generation: 1
  labels:
    app.kubernetes.io/component: application-controller
    app.kubernetes.io/instance: argocd
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: argocd-application-controller
    app.kubernetes.io/part-of: argocd
    app.kubernetes.io/version: v2.1.6
    helm.sh/chart: argo-cd-3.26.5
  managedFields:
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    manager: Go-http-client
    operation: Update
    time: "2021-11-09T21:40:57Z"
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:collisionCount: {}
        f:currentReplicas: {}
        f:currentRevision: {}
        f:observedGeneration: {}
        f:readyReplicas: {}
        f:replicas: {}
        f:updateRevision: {}
        f:updatedReplicas: {}
    manager: kube-controller-manager
    operation: Update
    time: "2021-11-09T21:41:48Z"
  name: argocd-application-controller
  namespace: argocd
  resourceVersion: "8313"
  uid: e4f894f4-e025-4868-a59b-4556bb50987c
spec:
  podManagementPolicy: OrderedReady
  replicas: 1
  revisionHistoryLimit: 5
  selector:
    matchLabels:
      app.kubernetes.io/instance: argocd
      app.kubernetes.io/name: argocd-application-controller
  serviceName: argocd-application-controller
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: application-controller
        app.kubernetes.io/instance: argocd
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: argocd-application-controller
        app.kubernetes.io/part-of: argocd
        app.kubernetes.io/version: v2.1.6
        helm.sh/chart: argo-cd-3.26.5
    spec:
      containers:
      - command:
        - argocd-application-controller
        - --status-processors
        - "20"
        - --operation-processors
        - "10"
        - --app-resync
        - "180"
        - --self-heal-timeout-seconds
        - "5"
        - --repo-server
        - argocd-repo-server:8081
        - --repo-server-timeout-seconds
        - "500"
        - --logformat
        - text
        - --loglevel
        - info
        - --redis
        - argocd-redis-ha-haproxy:6379
        - --repo-server-timeout-seconds
        - "500"
        image: quay.io/argoproj/argocd:v2.1.6
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 8082
            scheme: HTTP
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: application-controller
        ports:
        - containerPort: 8082
          name: controller
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          tcpSocket:
            port: 8082
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /home/argocd
          name: argocd-home
        - mountPath: /app/config/controller/tls
          name: argocd-repo-server-tls
        workingDir: /home/argocd
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: argocd-application-controller
      serviceAccountName: argocd-application-controller
      terminationGracePeriodSeconds: 30
      volumes:
      - emptyDir: {}
        name: argocd-home
      - name: argocd-repo-server-tls
        secret:
          defaultMode: 420
          items:
          - key: tls.crt
            path: tls.crt
          - key: tls.key
            path: tls.key
          - key: ca.crt
            path: ca.crt
          optional: true
          secretName: argocd-repo-server-tls
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
status:
  collisionCount: 0
  currentReplicas: 1
  currentRevision: argocd-application-controller-84b8c46c9b
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updateRevision: argocd-application-controller-84b8c46c9b
  updatedReplicas: 1
```

Still getting 
```
rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists
```

Sad 😢

**Comment by zonnie** [comment]
Actually scratch that @kshamajain99 - @gzur solution seems to do the job...
ArgoCD did become a lot slower in "converging" the state...but it's working now...nothing becomes "Unknown" due to the `helm dep` thing...Any idea when will this be actually solved ? (i.e. when is `helm` 3.7 is going to be merged to ArgoCD)

**Comment by KlavsKlavsen** [comment]
Helm 3.7.1 is merged in argocd master and should be part of next release (2.1.7) https://github.com/argoproj/argo-cd/commit/2770c690a5597fcbab344cd2ad494c918472bdd1

**Comment by KlavsKlavsen** [comment]
@zonnie only fix I've found "until 2.1.7 hopefullyn solves it with Helm 3.7.1" - is to simply KILL all argocd repo pods - and that will flush the cache which causes the problem.
The same way I have to kill all application-controller pods when sync hangs forever (happens with kube-prometheus chart f.ex.)

**Comment by gzur** [comment]
>  ...Any idea when will this be actually solved ? (i.e. when is `helm` 3.7 is going to be merged to ArgoCD)

The [fix](https://github.com/helm/helm/commit/4b23d0a25b15f990e5de9bc2e1e8cbd80de9243c) for the [underlying issue](https://github.com/helm/helm/pull/9889) was [released](https://github.com/helm/helm/releases/tag/v3.7.0) in `3.7.x` back September and has already been merged into Argo-CD:
 https://github.com/argoproj/argo-cd/blob/2770c690a5597fcbab344cd2ad494c918472bdd1/hack/tool-versions.sh#L12

Are you by any chance using `Helm 2` charts?

**Comment by zonnie** [comment]
The below example, AFAIK, is `helm` v3 correct ? 
`apiVersion: v2` is for `helm` 3 while `apiVersion: v1` is for `helm` 2 - correct ?
```
apiVersion: v2
name: secrets-app
description: App that contains secrets
type: application
version: 0.1.0
```

**Comment by zonnie** [comment]
@gzur ⬆️

**Comment by zonnie** [comment]
> @zonnie only fix I've found "until 2.1.7 hopefullyn solves it with Helm 3.7.1" - is to simply KILL all argocd repo pods - and that will flush the cache which causes the problem. The same way I have to kill all application-controller pods when sync hangs forever (happens with kube-prometheus chart f.ex.)

Doesn't work for me...it gets to the "crapped out" state pretty quickly - only @gzur solution mitigated the issue

**Comment by gzur** [comment]
> The below example, AFAIK, is `helm` v3 correct ? `apiVersion: v2` is for `helm` 3 while `apiVersion: v1` is for `helm` 2 - correct ?

Correct. 

I must admit that I am a bit surprised that https://github.com/helm/helm/pull/9889 did not resolve your issue. There were race-conditions in the previous code-path, which that fix has probably not addressed.

1. How many charts are you running?  
2. And do these charts have any subcharts?
3. How is the connectivity to the Helm Repository hosting these charts?



The reason I ask is that at my previous job - where we were experiencing this issue - we were downloading a metric ton of subcharts hosted by a Helm Repo that had connectivity issues - which was causing `helm dep update` to take a long time. 

We suspected that this was what exposed the aforementioned race condition.

**Comment by zonnie** [comment]
> > The below example, AFAIK, is `helm` v3 correct ? `apiVersion: v2` is for `helm` 3 while `apiVersion: v1` is for `helm` 2 - correct ?
> 
> 
> 
> Correct. 
> 
> 
> 
> I must admit that I am a bit surprised that https://github.com/helm/helm/pull/9889 did not resolve your issue. There was a possible race-condition in the previous code-path, which that fix has probably not addressed.
> 
> 
> 
> 1. How many charts are you running?  
> 
> 2. And do these charts have any subcharts?
> 
> 3. How is the connectivity to the Helm Repository hosting these charts?
> 
> 
> 
> 
> 
> 
> 
> The reason I ask is that at my previous job - where we were experiencing this issue - we were downloading a metric ton of subcharts hosted by a Helm Repo that had connectivity issues - which was causing `helm dep update` to take a long time. 
> 
> 
> 
> We suspected that this was what exposed the aforementioned race condition.
> 
> 
> 
> 

So
1. When u say running you mean the entire ArgoCD application count? If so we have them in the 600+ area. If u mean the specific creation act - it's around 10 at that specific moment which we create.
2. If by that u mean app-of-apps than we do - our "tree" has 3 top level applications, in 2 of those we have around 10-20 argo applications. The last one is a single argo application.
3. Our charts are of 2 types - in house charts are stored in GitHub in our repo. The 3rd parties are mostly Bitnami charts.

Thanks so much for your attention @gzur

**Comment by gzur** [comment]
> Thanks so much for your attention @gzur

Yeah, I don't understand why I'm so inordinately invested in this issue 😂  

One thing caught my eye though, @zonnie, [you wrote](https://github.com/argoproj/argo-cd/issues/5107#issuecomment-965265303):
> [...]
> Still getting
> 
> ```
> rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists
> ```
> 
> Sad 😢



That error string: `unable to move current charts to tmp dir` [got removed](https://github.com/helm/helm/commit/4b23d0a25b15f990e5de9bc2e1e8cbd80de9243c#diff-8157c4e80f3a8abb1dd88a83411502f7d02b91e112df1a6a7c25b176d3d18d71L262)  in the commit that addressed https://github.com/helm/helm/pull/9889 

So what version of ArgoCD are you ACTUALLY running?

**Comment by foxracle** [comment]
I have the exact same issue - with argocd 2.2.0.rc1
I have tried every solutions mentioned above, they does not work at all.

after downgrade to v2.1.7 yesterday, it seems like that it can syced quickly now but error message like "helm repo add charts.helm.sh https://charts.helm.sh/stable` failed" appears once after downgrade.

all other error messages likes 'context deadline exceeded' or 'helm dependency build" failure' gone.

**Comment by zonnie** [comment]
@foxracle I have to say that @gzur solution worked for me.
I have quite a lot of Argo Applications (hundreds)...it used to be unusable, now I have no issues

**Comment by foxracle** [comment]
@zonnie I have tried these, I also try to flush all redis cache, but  nothing happened. I do not think it is a problem of load or resources lack of argo-repo-server or something else. 

```
argocd:
  controller:
    extraArgs:
      - --repo-server-timeout-seconds
      - "500" 
  repoServer:
    env:
      - name: "ARGOCD_EXEC_TIMEOUT"
        value: "5m"

```

In fact, I have installed two argocd services in two different k8s cluster, one is v1.7.8 and manages two k8s cluster, one is v2.2.0.rc1 and manages one k8s cluster. they are all use the same git repo with 50+ apps to sync.  the v1.7.8 is ok with no args tuning, but v2.2.0.rc1 is not.  after I tried every solutions I googled, I give up by downgrading to v2.1.7 stable version

**Comment by qtheya** [comment]
> Seeing as this issue is "resolved" temporarily - by killing repo-server pods, so they get recreated - its clearly a caching problem.
repo-server and `redis` pods ...

**Comment by emretanriverdi** [comment]
@gzur's solution worked for me.

**Comment by sdelrio** [comment]
Yes, I'm using helm chart to deploy ArgoCD. Using last chart I saw 3.33.5 (5 Feb 2022)
Using this chart version and rolling back to v2.2.2 also fix the issue. The previous chart version I was using was 3.29.5

**Comment by begemotik** [comment]
> Yes, I'm using helm chart to deploy ArgoCD. Using last chart I saw 3.33.5 (5 Feb 2022)
> Using this chart version and rolling back to v2.2.2 also fix the issue. The previous chart version I was using was 3.29.5
> 

Just try out the 3.29.5 and specify image tag 2.2.5, it will work fine. We were trying to identify the problem with chart causing that and we have some clues, but no confirmation so far

**Comment by sdelrio** [comment]
> > Yes, I'm using helm chart to deploy ArgoCD. Using last chart I saw 3.33.5 (5 Feb 2022)
> > Using this chart version and rolling back to v2.2.2 also fix the issue. The previous chart version I was using was 3.29.5
> 
> Just try out the 3.29.5 and specify image tag 2.2.5, it will work fine. We were trying to identify the problem with chart causing that and we have some clues, but no confirmation so far

Thank you very much. I will test 2.2.5 + chart 3.29.5.
For what I just saw the main differences are `Poddisruptionbudget` and initcontainer `copyutil` for copy argocd binary, but they seems to be disabled by default.

**Comment by willemm** [comment]
We're having similar issues.

One thing I would like to add (which is why I'm commenting) is that I believe an application should never get into a permanently broken state just because of a temporary issue like this.

i.e. if I click 'hard refresh' on the applications that are broken like this, they become fixed.  I don't want to have to do that manually, it should just sort itself out.

**Comment by tomikonio** [comment]
Hi,
Long thread here.
Is there a TLDR - can I tell ArgoCD to ignore helm chart dependencies so that the sync will apply changes to other components?

**Comment by foxracle** [comment]
these  are the two terrible problems after working with argocd two years from version 1.7.x ~ 2.3.x, especially in an urgent deploy case.

- ComparisonError: rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed timeout after 1m30s
- ComparisonError: rpc error: code = DeadlineExceeded desc = context deadline exceeded

most time, I know it is not a performance issue after I followed everything from the official doc[[high availability](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/)], it is just a cache problem.
during these 2 years, there is still only a workaround to fix it：recreate argocd-repo-server and flush all data in redis cluster.  thanks God!

**Comment by crenshaw-dev** [comment]
Quick note on this one:

> ComparisonError: rpc error: code = DeadlineExceeded desc = context deadline exceeded

That's a generic error message from the golang `context` package, and it just means "a timeout happened somewhere."

We've made efforts recently to always wrap all error messages to provide more context. Hopefully in future versions the _reason_ for the timeout will be much more clear.

**Comment by chenshap** [comment]
hello, did anyone find a solution? we are experiencing  this on argocd version 5.16.1

**Comment by ArieLevs** [comment]
+1 here, still happens with v2.6.1
constant `"helm dependency build" failure` errors
cannot see any pressure with repo server

**Comment by zengzhengrong** [comment]
I have the same , It mean delete  Chart.lock file of git repo?
```
rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: the lock file (Chart.lock) is out of sync with the dependencies file (Chart.yaml). Please update the dependencies
````

**Comment by slntopp** [comment]
Had same issue tho it was completely unrelated to Helm dependencies
Actual issue was CronJob template which had typo in `Kind`
